### DISCUSSION TITLE
How use zustand on a server

### DISCUSSION THREAD

**Discussion by nicitaacom** [pmndrs/zustand#2124] [Q&A]
I use Next 13.5 + TypeScript + zustand

I know that localstorage exist only on a client

Is it possible to avoid this error message?
```
Error: useRef only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/react-client-hook-in-server-component
```

getAnonymousCart.ts
```ts
import useAnonymousCartStore from "@/store/user/anonymousCart"

export function getAnonymousCart() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const anonymousCart = useAnonymousCartStore()
  return anonymousCart.cartProducts
}
```

Products.tsx
```tsx

async function fetchCartProducts() {
  //IN PROGRESS - create case for unauthenticated user I mean if !user output something
  const { data: user } = await supabaseServer().auth.getUser()
  if (user?.user?.id) {
    //if user - output quantity for products based on users_cart
    const cartProducts = await supabaseServer()
      .from("users_cart")
      .select("cart_products")
      .eq("id", user.user.id)
      .single()
    return cartProducts.data?.cart_products as unknown as ICartProduct[]
  } else {
    //if !user - output quantity for products based on localstorage
    return getAnonymousCart()
  }
}
```

I want render products quantity
if (!user) - renader product.quantity based on localstorage
if (user) - render product.quantity based on DB

I know it possible to store info in cookies somehow but I tried and when I tried to update state in cookies I got string (I used JSON.parse)

So once again - its some approach to keep performance and use SSR?
Or if !user I get worse performance because I use 'use client'?

**Comment by dai-shi** [comment]
I'm not sure if Zustand behaves as you want on server, but you could read your store just like a module variable.

```js
export function getAnonymousCart() {
  const anonymousCart = useAnonymousCartStore.getState()
  return anonymousCart.cartProducts
}
```

But it doesn't read from localStorage.

> //if !user - output quantity for products based on localstorage

You would never be able to read localStorage on server, as you said.


See also: https://github.com/pmndrs/zustand/discussions/2122#discussioncomment-7289916
